<a href="https://colab.research.google.com/github/Pranaybannu/PLAY-STORE-APP-DATA-EDA/blob/main/PLAYSTORE.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Project Name**    -



##### **Project Type**    - EDA FOR PLAYSTORE APP DATA
##### **Contribution**    - Individual(Pranay raj sambodhu)

# **Project Summary -**

The **Play store app data** is a **major boon** for app making buisinesses.This platform allows the app users to write reviews,comments and also allows the users to score/rate the app based on their experience with the app.

In this project, I'm going to perform EDA on the playstore app data. This EDA involves understanding the distribution of app category,installs,pricing,user reviews,comments and ratings. It also involves the handling any missing,duplicate values and other data inconsistencies. This is followed by plotting visualization graphs using plotly and seaborn for better client understanding.This allows the client/viewers to understand how app categories,installs, pricing,type and size influence user reviews,rating and comments.This EDA also involves analysing the user reviews, wheater the reviews/comments are mostly subjective or objective.

The given dataset is divided into two parts/files:
1. **Playstore data.csv:**
*  This dataset contains app information with columns liek app category,type,app size,app pricing,no.of installs etc., with 10k+ rows.

2. **User reviews data.csv**
*  This dataset contains columns like app name, user reviews,sentiment, sentiment polarity and sentiment subjectivity with 64k+ rows.

EDA will be performed on both of these files and buisiness insights,positives and negatives will be extarcted in the following cells.

# **GitHub Link -**

Provide your GitHub Link here.

# ***Let's Begin !***

## ***1. Know Your Data***

### Import Libraries

In [ ]:
# Import Libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px
from google.colab import files
import plotly.graph_objects as go

### Dataset Loading

In [ ]:
# Load Dataset
pd.set_option('display.max_rows',None)
pd.set_option('display.max_columns',None)

app = pd.read_csv('/content/drive/MyDrive/PLAYSTORE/Play Store Data.csv')
rev = pd.read_csv('/content/drive/MyDrive/PLAYSTORE/User Reviews.csv')

### Dataset First View

In [ ]:
# Dataset First Look
app.head()  ## for playstore app data

,App,Category,Rating,Reviews,Size,Installs,Type,Price,Content Rating,Genres,Last Updated,Current Ver,Android Ver
0,Photo Editor & Candy Camera & Grid & ScrapBook,ART_AND_DESIGN,4.1,159,19M,"10,000+",Free,0,Everyone,Art & Design,"January 7, 2018",1.0.0,4.0.3 and up
1,Coloring book moana,ART_AND_DESIGN,3.9,967,14M,"500,000+",Free,0,Everyone,Art & Design;Pretend Play,"January 15, 2018",2.0.0,4.0.3 and up
2,"U Launcher Lite – FREE Live Cool Themes, Hide ...",ART_AND_DESIGN,4.7,87510,8.7M,"5,000,000+",Free,0,Everyone,Art & Design,"August 1, 2018",1.2.4,4.0.3 and up
3,Sketch - Draw & Paint,ART_AND_DESIGN,4.5,215644,25M,"50,000,000+",Free,0,Teen,Art & Design,"June 8, 2018",Varies with device,4.2 and up
4,Pixel Draw - Number Art Coloring Book,ART_AND_DESIGN,4.3,967,2.8M,"100,000+",Free,0,Everyone,Art & Design;Creativity,"June 20, 2018",1.1,4.4 and up


In [ ]:
rev.head() ##for user review data

,App,Translated_Review,Sentiment,Sentiment_Polarity,Sentiment_Subjectivity
0,10 Best Foods for You,I like eat delicious food. That's I'm cooking ...,Positive,1.00,0.533333
1,10 Best Foods for You,This help eating healthy exercise regular basis,Positive,0.25,0.288462
2,10 Best Foods for You,NaN,NaN,NaN,NaN
3,10 Best Foods for You,Works great especially going grocery store,Positive,0.40,0.875000
4,10 Best Foods for You,Best idea us,Positive,1.00,0.300000


### Dataset Rows & Columns count

In [ ]:
# Dataset Rows & Columns count
app.shape,rev.shape

((10841, 13), (64295, 5))

### Dataset Information

In [ ]:
# Dataset Info
app.info()  ##for playstore data

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10841 entries, 0 to 10840
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   App             10841 non-null  object 
 1   Category        10841 non-null  object 
 2   Rating          9367 non-null   float64
 3   Reviews         10841 non-null  object 
 4   Size            10841 non-null  object 
 5   Installs        10841 non-null  object 
 6   Type            10840 non-null  object 
 7   Price           10841 non-null  object 
 8   Content Rating  10840 non-null  object 
 9   Genres          10841 non-null  object 
 10  Last Updated    10841 non-null  object 
 11  Current Ver     10833 non-null  object 
 12  Android Ver     10838 non-null  object 
dtypes: float64(1), object(12)
memory usage: 1.1+ MB


In [ ]:
##info of user reviews data
rev.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 64295 entries, 0 to 64294
Data columns (total 5 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   App                     64295 non-null  object 
 1   Translated_Review       37427 non-null  object 
 2   Sentiment               37432 non-null  object 
 3   Sentiment_Polarity      37432 non-null  float64
 4   Sentiment_Subjectivity  37432 non-null  float64
dtypes: float64(2), object(3)
memory usage: 2.5+ MB


#### Duplicate Values

In [ ]:
# Dataset Duplicate Value Count
app.duplicated().mean()*100        ##playstore has 4.4% duplicated rows

np.float64(4.455308550871691)

In [ ]:
##identifying the duplicates of playstore app data
app[app.duplicated].head(10)

,App,Category,Rating,Reviews,Size,Installs,Type,Price,Content Rating,Genres,Last Updated,Current Ver,Android Ver
229,Quick PDF Scanner + OCR FREE,BUSINESS,4.2,80805,Varies with device,"5,000,000+",Free,0,Everyone,Business,"February 26, 2018",Varies with device,4.0.3 and up
236,Box,BUSINESS,4.2,159872,Varies with device,"10,000,000+",Free,0,Everyone,Business,"July 31, 2018",Varies with device,Varies with device
239,Google My Business,BUSINESS,4.4,70991,Varies with device,"5,000,000+",Free,0,Everyone,Business,"July 24, 2018",2.19.0.204537701,4.4 and up
256,ZOOM Cloud Meetings,BUSINESS,4.4,31614,37M,"10,000,000+",Free,0,Everyone,Business,"July 20, 2018",4.1.28165.0716,4.0 and up
261,join.me - Simple Meetings,BUSINESS,4.0,6989,Varies with device,"1,000,000+",Free,0,Everyone,Business,"July 16, 2018",4.3.0.508,4.4 and up
265,Box,BUSINESS,4.2,159872,Varies with device,"10,000,000+",Free,0,Everyone,Business,"July 31, 2018",Varies with device,Varies with device
266,Zenefits,BUSINESS,4.2,296,14M,"50,000+",Free,0,Everyone,Business,"June 15, 2018",3.2.1,4.1 and up
267,Google Ads,BUSINESS,4.3,29313,20M,"5,000,000+",Free,0,Everyone,Business,"July 30, 2018",1.12.0,4.0.3 and up
268,Google My Business,BUSINESS,4.4,70991,Varies with device,"5,000,000+",Free,0,Everyone,Business,"July 24, 2018",2.19.0.204537701,4.4 and up
269,Slack,BUSINESS,4.4,51507,Varies with device,"5,000,000+",Free,0,Everyone,Business,"August 2, 2018",Varies with device,Varies with device


In [ ]:
##duplicates of user reviews data
rev.duplicated().mean()*100    #52% of user reviews data is duplicated

np.float64(52.284003421727974)

In [ ]:
rev[rev.duplicated()].head(10)

,App,Translated_Review,Sentiment,Sentiment_Polarity,Sentiment_Subjectivity
7,10 Best Foods for You,NaN,NaN,NaN,NaN
15,10 Best Foods for You,NaN,NaN,NaN,NaN
78,10 Best Foods for You,Good,Positive,0.70,0.600000
79,10 Best Foods for You,Good,Positive,0.70,0.600000
100,10 Best Foods for You,I like eat delicious food. That's I'm cooking ...,Positive,1.00,0.533333
101,10 Best Foods for You,This help eating healthy exercise regular basis,Positive,0.25,0.288462
102,10 Best Foods for You,NaN,NaN,NaN,NaN
103,10 Best Foods for You,Works great especially going grocery store,Positive,0.40,0.875000
104,10 Best Foods for You,Best idea us,Positive,1.00,0.300000
105,10 Best Foods for You,Best way,Positive,1.00,0.300000


#### Missing Values/Null Values

In [ ]:
# Missing Values/Null Values Count in playstore data
app.isnull().mean()*100  ## there are nullvalues in the playstore app data

,0
App,0.000000
Category,0.000000
Rating,13.596532
Reviews,0.000000
Size,0.000000
Installs,0.000000
Type,0.009224
Price,0.000000
Content Rating,0.009224
Genres,0.000000


In [ ]:
# Visualizing the missing values in playstore app data
app[app['Rating'].isnull()].head(10)

,App,Category,Rating,Reviews,Size,Installs,Type,Price,Content Rating,Genres,Last Updated,Current Ver,Android Ver
23,Mcqueen Coloring pages,ART_AND_DESIGN,NaN,61,7.0M,"100,000+",Free,0,Everyone,Art & Design;Action & Adventure,"March 7, 2018",1.0.0,4.1 and up
113,Wrinkles and rejuvenation,BEAUTY,NaN,182,5.7M,"100,000+",Free,0,Everyone 10+,Beauty,"September 20, 2017",8.0,3.0 and up
123,Manicure - nail design,BEAUTY,NaN,119,3.7M,"50,000+",Free,0,Everyone,Beauty,"July 23, 2018",1.3,4.1 and up
126,Skin Care and Natural Beauty,BEAUTY,NaN,654,7.4M,"100,000+",Free,0,Teen,Beauty,"July 17, 2018",1.15,4.1 and up
129,"Secrets of beauty, youth and health",BEAUTY,NaN,77,2.9M,"10,000+",Free,0,Mature 17+,Beauty,"August 8, 2017",2.0,2.3 and up
130,Recipes and tips for losing weight,BEAUTY,NaN,35,3.1M,"10,000+",Free,0,Everyone 10+,Beauty,"December 11, 2017",2.0,3.0 and up
134,"Lady adviser (beauty, health)",BEAUTY,NaN,30,9.9M,"10,000+",Free,0,Mature 17+,Beauty,"January 24, 2018",3.0,3.0 and up
163,Anonymous caller detection,BOOKS_AND_REFERENCE,NaN,161,2.7M,"10,000+",Free,0,Everyone,Books & Reference,"July 13, 2018",1.0,2.3 and up
180,SH-02J Owner's Manual (Android 8.0),BOOKS_AND_REFERENCE,NaN,2,7.2M,"50,000+",Free,0,Everyone,Books & Reference,"June 15, 2018",3.0,6.0 and up
185,URBANO V 02 instruction manual,BOOKS_AND_REFERENCE,NaN,114,7.3M,"100,000+",Free,0,Everyone,Books & Reference,"August 7, 2015",1.1,5.1 and up


In [ ]:
# Missing Values/Null Values Count in user reviews data
rev.isnull().mean()*100   ### too many null values

,0
App,0.000000
Translated_Review,41.788631
Sentiment,41.780854
Sentiment_Polarity,41.780854
Sentiment_Subjectivity,41.780854


In [ ]:
# Visualizing the missing values in user review data
rev[rev['Translated_Review'].isnull()].head(10)

,App,Translated_Review,Sentiment,Sentiment_Polarity,Sentiment_Subjectivity
2,10 Best Foods for You,NaN,NaN,NaN,NaN
7,10 Best Foods for You,NaN,NaN,NaN,NaN
15,10 Best Foods for You,NaN,NaN,NaN,NaN
102,10 Best Foods for You,NaN,NaN,NaN,NaN
107,10 Best Foods for You,NaN,NaN,NaN,NaN
115,10 Best Foods for You,NaN,NaN,NaN,NaN
268,11st,NaN,Neutral,0.0,0.0
362,1LINE – One Line with One Touch,NaN,NaN,NaN,NaN
368,1LINE – One Line with One Touch,NaN,NaN,NaN,NaN
405,2018Emoji Keyboard 😂 Emoticons Lite -sticker&gif,NaN,NaN,NaN,NaN


### What did you know about your dataset?

* **App data** has 10k+rows, 4.45% duplicated data and columns like ratings,content rating and version contain null values.
* **reviews data** has 64k+ rows, 54% duplicated data and columns like translated reviews,sentiment and sentiment subjectivity has null values(42%0

## ***2. Understanding Your Variables***

In [ ]:
# Dataset Columns of playstore app data
app.columns

Index(['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type',
       'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver',
       'Android Ver'],
      dtype='object')

In [ ]:
rev.columns  ##columns of reviews data

Index(['App', 'Translated_Review', 'Sentiment', 'Sentiment_Polarity',
       'Sentiment_Subjectivity'],
      dtype='object')

In [ ]:
# Dataset Describe for playstore data
app.describe(include='object')

,App,Category,Reviews,Size,Installs,Type,Price,Content Rating,Genres,Last Updated,Current Ver,Android Ver
count,10841,10841,10841,10841,10841,10840,10841,10840,10841,10841,10833,10838
unique,9660,34,6002,462,22,3,93,6,120,1378,2832,33
top,ROBLOX,FAMILY,0,Varies with device,"1,000,000+",Free,0,Everyone,Tools,"August 3, 2018",Varies with device,4.1 and up
freq,9,1972,596,1695,1579,10039,10040,8714,842,326,1459,2451


In [ ]:
app.describe(include='number')

,Rating
count,9367.000000
mean,4.193338
std,0.537431
min,1.000000
25%,4.000000
50%,4.300000
75%,4.500000
max,19.000000


In [ ]:
# Dataset Describe for users data
rev.describe(include='object')

,App,Translated_Review,Sentiment
count,64295,37427,37432
unique,1074,27994,3
top,Bowmasters,Good,Positive
freq,320,247,23998


In [ ]:
rev.describe(include='number')

,Sentiment_Polarity,Sentiment_Subjectivity
count,37432.000000,37432.000000
mean,0.182146,0.492704
std,0.351301,0.259949
min,-1.000000,0.000000
25%,0.000000,0.357143
50%,0.150000,0.514286
75%,0.400000,0.650000
max,1.000000,1.000000


In [ ]:
app.head()

,App,Category,Rating,Reviews,Size,Installs,Type,Price,Content Rating,Genres,Last Updated,Current Ver,Android Ver
0,Photo Editor & Candy Camera & Grid & ScrapBook,ART_AND_DESIGN,4.1,159,19M,"10,000+",Free,0,Everyone,Art & Design,"January 7, 2018",1.0.0,4.0.3 and up
1,Coloring book moana,ART_AND_DESIGN,3.9,967,14M,"500,000+",Free,0,Everyone,Art & Design;Pretend Play,"January 15, 2018",2.0.0,4.0.3 and up
2,"U Launcher Lite – FREE Live Cool Themes, Hide ...",ART_AND_DESIGN,4.7,87510,8.7M,"5,000,000+",Free,0,Everyone,Art & Design,"August 1, 2018",1.2.4,4.0.3 and up
3,Sketch - Draw & Paint,ART_AND_DESIGN,4.5,215644,25M,"50,000,000+",Free,0,Teen,Art & Design,"June 8, 2018",Varies with device,4.2 and up
4,Pixel Draw - Number Art Coloring Book,ART_AND_DESIGN,4.3,967,2.8M,"100,000+",Free,0,Everyone,Art & Design;Creativity,"June 20, 2018",1.1,4.4 and up


In [ ]:
rev.head()

,App,Translated_Review,Sentiment,Sentiment_Polarity,Sentiment_Subjectivity
0,10 Best Foods for You,I like eat delicious food. That's I'm cooking ...,Positive,1.00,0.533333
1,10 Best Foods for You,This help eating healthy exercise regular basis,Positive,0.25,0.288462
2,10 Best Foods for You,NaN,NaN,NaN,NaN
3,10 Best Foods for You,Works great especially going grocery store,Positive,0.40,0.875000
4,10 Best Foods for You,Best idea us,Positive,1.00,0.300000


### Variables Description

**PLAYSTORE APP DATA**:
* App--> **Photo Editor & Candy Camera & Grid & ScrapBook (string) --> App name**
*	Category--> **ART_AND_DESIGN (string) --> To which catgeory the app belongs to**
* Rating--> **4.1 (string) --> from 1-5**
* Reviews--> **159 (float) -- Total no.of reviews**
* Size--> **19M (string) --> size of the app in MB**
*	Installs--> **10,000+ (string) --> total no of times the app is installed**
* Type--> **Free (string) --> free or priced**
* Price	--> **0 (string) --> 0 if free or else number in string**
* Content Rating--> **Everyone (string) --> content age category rating**
* Genres--> **Art & Design (string) --> genre type**
* Last Updated--> **January 7 (string) --> recent app updated date**
* Current Ver--> **2018	1.0.0 (string) --> what is the current version**
* Android Ver--> **4.0.3 and up (string) --> typical android version**

**USER REVIEW DATA:**
* App	--> **10 Best Foods for You	(string) --> app name**
* Translated_Review	--> **I like eat delicious food. That's I'm cooking ...(string) ---> review translated to english**
* Sentiment	--> **Positive (string) --> type buisiness positive or negative.**
* Sentiment_Polarity --> **1.00 (float) --> strength of sentiment (strong positive or negative)**
* Sentiment_Subjectivity --> **0.533333 (float) (subjective or objective review)**


### Check Unique Values for each variable.

**Check Unique Values for each variable in playstore data**

In [ ]:
# Check Unique Values for each variable.
## app category
app['Category'].value_counts()

,count
Category,
FAMILY,1972
GAME,1144
TOOLS,843
MEDICAL,463
BUSINESS,460
PRODUCTIVITY,424
PERSONALIZATION,392
COMMUNICATION,387
SPORTS,384


In [ ]:
##rating
app['Rating'].value_counts()

,count
Rating,
4.4,1109
4.3,1076
4.5,1038
4.2,952
4.6,823
4.1,708
4.0,568
4.7,499
3.9,386


In [ ]:
##size
app['Size'].value_counts().head()

,count
Size,
Varies with device,1695
11M,198
12M,196
14M,194
13M,191


In [ ]:
## free apps
app['Type'].value_counts()

,count
Type,
Free,10039
Paid,800
0,1


In [ ]:
##Content Rating
app['Content Rating'].value_counts()

,count
Content Rating,
Everyone,8714
Teen,1208
Mature 17+,499
Everyone 10+,414
Adults only 18+,3
Unrated,2


In [ ]:
##Genres
app['Genres'].value_counts().head(10)

,count
Genres,
Tools,842
Entertainment,623
Education,549
Medical,463
Business,460
Productivity,424
Sports,398
Personalization,392
Communication,387


**Check Unique Values for each variable in User reviews data**

In [ ]:
rev.head()

,App,Translated_Review,Sentiment,Sentiment_Polarity,Sentiment_Subjectivity
0,10 Best Foods for You,I like eat delicious food. That's I'm cooking ...,Positive,1.00,0.533333
1,10 Best Foods for You,This help eating healthy exercise regular basis,Positive,0.25,0.288462
2,10 Best Foods for You,NaN,NaN,NaN,NaN
3,10 Best Foods for You,Works great especially going grocery store,Positive,0.40,0.875000
4,10 Best Foods for You,Best idea us,Positive,1.00,0.300000


In [ ]:
## sentiment
rev['Sentiment'].value_counts().head()

,count
Sentiment,
Positive,23998
Negative,8271
Neutral,5163


## 3. ***Data Wrangling***

### Data Wrangling Code for Play store app data

In [ ]:
# checking duplicates
app.duplicated().sum()

np.int64(483)

In [ ]:
app.drop_duplicates(inplace=True) ##dropping all duplicates

In [ ]:
app.duplicated().sum() ##zero duplicated rows

np.int64(0)

In [ ]:
app.info()

<class 'pandas.core.frame.DataFrame'>
Index: 10358 entries, 0 to 10840
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   App             10358 non-null  object 
 1   Category        10358 non-null  object 
 2   Rating          8893 non-null   float64
 3   Reviews         10358 non-null  object 
 4   Size            10358 non-null  object 
 5   Installs        10358 non-null  object 
 6   Type            10357 non-null  object 
 7   Price           10358 non-null  object 
 8   Content Rating  10357 non-null  object 
 9   Genres          10358 non-null  object 
 10  Last Updated    10358 non-null  object 
 11  Current Ver     10350 non-null  object 
 12  Android Ver     10355 non-null  object 
dtypes: float64(1), object(12)
memory usage: 1.1+ MB


In [ ]:
##handling missing values
spl = ['-','no','null','unknown',' ','miss','missing']
app.replace(spl,np.nan,inplace=True)

In [ ]:
app.info() ##no spl characters

<class 'pandas.core.frame.DataFrame'>
Index: 10358 entries, 0 to 10840
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   App             10358 non-null  object 
 1   Category        10358 non-null  object 
 2   Rating          8893 non-null   float64
 3   Reviews         10358 non-null  object 
 4   Size            10358 non-null  object 
 5   Installs        10358 non-null  object 
 6   Type            10357 non-null  object 
 7   Price           10358 non-null  object 
 8   Content Rating  10357 non-null  object 
 9   Genres          10358 non-null  object 
 10  Last Updated    10358 non-null  object 
 11  Current Ver     10350 non-null  object 
 12  Android Ver     10355 non-null  object 
dtypes: float64(1), object(12)
memory usage: 1.1+ MB


In [ ]:
app['Rating'].isnull().mean()*100

np.float64(14.143657076655725)

In [ ]:
app.loc[app['Rating'].isna(),'Last Updated'].head(10)  ## keeping them as null to maintain data orginality

,Last Updated
23,"March 7, 2018"
113,"September 20, 2017"
123,"July 23, 2018"
126,"July 17, 2018"
129,"August 8, 2017"
130,"December 11, 2017"
134,"January 24, 2018"
163,"July 13, 2018"
180,"June 15, 2018"
185,"August 7, 2015"


**VARIABLE ENGINEERING**

In [ ]:
### making a new column by converting  last update column to datetime

In [ ]:
app[app['Last Updated']=='1.0.19']

,App,Category,Rating,Reviews,Size,Installs,Type,Price,Content Rating,Genres,Last Updated,Current Ver,Android Ver
10472,Life Made WI-Fi Touchscreen Photo Frame,1.9,19.0,3.0M,"1,000+",Free,0,Everyone,NaN,"February 11, 2018",1.0.19,4.0 and up,NaN


In [ ]:
##adjusting the row and column values

app.loc[
    app['App'] == 'Life Made WI-Fi Touchscreen Photo Frame',
    'Category']=np.nan

app.loc[
    app['App'] == 'Life Made WI-Fi Touchscreen Photo Frame', 'Rating']=1.9

app.loc[
    app['App'] == 'Life Made WI-Fi Touchscreen Photo Frame','Reviews'] = '19.0'

app.loc[
    app['App'] == 'Life Made WI-Fi Touchscreen Photo Frame','Size']='3.0M'

app.loc[
    app['App'] == 'Life Made WI-Fi Touchscreen Photo Frame','Installs']='1,000+'

app.loc[
    app['App'] == 'Life Made WI-Fi Touchscreen Photo Frame','Type']='Free'

app.loc[
    app['App'] == 'Life Made WI-Fi Touchscreen Photo Frame','Price']='0.0'

app.loc[
    app['App'] == 'Life Made WI-Fi Touchscreen Photo Frame','Content Rating']='Everyone'

app.loc[
    app['App'] == 'Life Made WI-Fi Touchscreen Photo Frame','Genres']=np.nan

app.loc[
    app['App'] == 'Life Made WI-Fi Touchscreen Photo Frame','Last Updated']='February 11, 2018'

app.loc[
    app['App'] == 'Life Made WI-Fi Touchscreen Photo Frame','Current Ver']='1.0.19'

app.loc[
    app['App'] == 'Life Made WI-Fi Touchscreen Photo Frame','Android Ver']='4.0 and up'


In [ ]:
app[app['Last Updated']=='February 11, 2018']

,App,Category,Rating,Reviews,Size,Installs,Type,Price,Content Rating,Genres,Last Updated,Current Ver,Android Ver
2903,"High-Speed Camera (GIF,Burst)",PHOTOGRAPHY,4.1,78140,10M,"5,000,000+",Free,0,Everyone,Photography,"February 11, 2018",5.5.2,4.0.3 and up
5299,AK 47 Lock Screen,PERSONALIZATION,4.6,5,8.8M,500+,Free,0,Mature 17+,Personalization,"February 11, 2018",1.0,4.0.3 and up
6337,BJ's Community SoundBoard,FAMILY,NaN,0,Varies with device,5+,Free,0,Teen,Entertainment,"February 11, 2018",Varies with device,Varies with device
8887,The Secret Daily Teachings,LIFESTYLE,4.3,206,32M,"1,000+",Paid,$4.99,Everyone,Lifestyle,"February 11, 2018",14.0,4.0.3 and up
9188,EB Scanner,PRODUCTIVITY,5.0,9,25M,50+,Free,0,Everyone,Productivity,"February 11, 2018",1.7.8,4.0.3 and up
9502,Pyaar Ek Dhoka hai Game - Anti Valentines,FAMILY,NaN,8,17M,100+,Free,0,Everyone,Casual,"February 11, 2018",3,4.1 and up
10472,Life Made WI-Fi Touchscreen Photo Frame,NaN,1.9,19.0,3.0M,"1,000+",Free,0.0,Everyone,NaN,"February 11, 2018",1.0.19,4.0 and up


In [ ]:
app['updated'] = pd.to_datetime(app['Last Updated'],format='mixed')

In [ ]:
app.head()

,App,Category,Rating,Reviews,Size,Installs,Type,Price,Content Rating,Genres,Last Updated,Current Ver,Android Ver,updated
0,Photo Editor & Candy Camera & Grid & ScrapBook,ART_AND_DESIGN,4.1,159,19M,"10,000+",Free,0,Everyone,Art & Design,"January 7, 2018",1.0.0,4.0.3 and up,2018-01-07
1,Coloring book moana,ART_AND_DESIGN,3.9,967,14M,"500,000+",Free,0,Everyone,Art & Design;Pretend Play,"January 15, 2018",2.0.0,4.0.3 and up,2018-01-15
2,"U Launcher Lite – FREE Live Cool Themes, Hide ...",ART_AND_DESIGN,4.7,87510,8.7M,"5,000,000+",Free,0,Everyone,Art & Design,"August 1, 2018",1.2.4,4.0.3 and up,2018-08-01
3,Sketch - Draw & Paint,ART_AND_DESIGN,4.5,215644,25M,"50,000,000+",Free,0,Teen,Art & Design,"June 8, 2018",Varies with device,4.2 and up,2018-06-08
4,Pixel Draw - Number Art Coloring Book,ART_AND_DESIGN,4.3,967,2.8M,"100,000+",Free,0,Everyone,Art & Design;Creativity,"June 20, 2018",1.1,4.4 and up,2018-06-20


In [ ]:
## extracting year and month from updated column
app['year']= app['updated'].dt.year


In [ ]:
app['month']=app['updated'].dt.month_name()

In [ ]:
app.head()

,App,Category,Rating,Reviews,Size,Installs,Type,Price,Content Rating,Genres,Last Updated,Current Ver,Android Ver,updated,year,month
0,Photo Editor & Candy Camera & Grid & ScrapBook,ART_AND_DESIGN,4.1,159,19M,"10,000+",Free,0,Everyone,Art & Design,"January 7, 2018",1.0.0,4.0.3 and up,2018-01-07,2018,January
1,Coloring book moana,ART_AND_DESIGN,3.9,967,14M,"500,000+",Free,0,Everyone,Art & Design;Pretend Play,"January 15, 2018",2.0.0,4.0.3 and up,2018-01-15,2018,January
2,"U Launcher Lite – FREE Live Cool Themes, Hide ...",ART_AND_DESIGN,4.7,87510,8.7M,"5,000,000+",Free,0,Everyone,Art & Design,"August 1, 2018",1.2.4,4.0.3 and up,2018-08-01,2018,August
3,Sketch - Draw & Paint,ART_AND_DESIGN,4.5,215644,25M,"50,000,000+",Free,0,Teen,Art & Design,"June 8, 2018",Varies with device,4.2 and up,2018-06-08,2018,June
4,Pixel Draw - Number Art Coloring Book,ART_AND_DESIGN,4.3,967,2.8M,"100,000+",Free,0,Everyone,Art & Design;Creativity,"June 20, 2018",1.1,4.4 and up,2018-06-20,2018,June


In [ ]:
app['Genres'].unique()

array(['Art & Design', 'Art & Design;Pretend Play',
       'Art & Design;Creativity', 'Art & Design;Action & Adventure',
       'Auto & Vehicles', 'Beauty', 'Books & Reference', 'Business',
       'Comics', 'Comics;Creativity', 'Communication', 'Dating',
       'Education;Education', 'Education', 'Education;Creativity',
       'Education;Music & Video', 'Education;Action & Adventure',
       'Education;Pretend Play', 'Education;Brain Games', 'Entertainment',
       'Entertainment;Music & Video', 'Entertainment;Brain Games',
       'Entertainment;Creativity', 'Events', 'Finance', 'Food & Drink',
       'Health & Fitness', 'House & Home', 'Libraries & Demo',
       'Lifestyle', 'Lifestyle;Pretend Play',
       'Adventure;Action & Adventure', 'Arcade', 'Casual', 'Card',
       'Casual;Pretend Play', 'Action', 'Strategy', 'Puzzle', 'Sports',
       'Music', 'Word', 'Racing', 'Casual;Creativity',
       'Casual;Action & Adventure', 'Simulation', 'Adventure', 'Board',
       'Trivia', 'Role 

In [ ]:
def prim_gen(x):
    if not isinstance(x, str):
        return x                                       ##for primary genres
    a = x.find(';')
    if a != -1:
        return x[:a]
    return x


In [ ]:
##creating primary genre column
app['primary_genre']= app['Genres'].apply(prim_gen)

In [ ]:
app.head()

,App,Category,Rating,Reviews,Size,Installs,Type,Price,Content Rating,Genres,Last Updated,Current Ver,Android Ver,updated,year,month,primary_genre
0,Photo Editor & Candy Camera & Grid & ScrapBook,ART_AND_DESIGN,4.1,159,19M,"10,000+",Free,0,Everyone,Art & Design,"January 7, 2018",1.0.0,4.0.3 and up,2018-01-07,2018,January,Art & Design
1,Coloring book moana,ART_AND_DESIGN,3.9,967,14M,"500,000+",Free,0,Everyone,Art & Design;Pretend Play,"January 15, 2018",2.0.0,4.0.3 and up,2018-01-15,2018,January,Art & Design
2,"U Launcher Lite – FREE Live Cool Themes, Hide ...",ART_AND_DESIGN,4.7,87510,8.7M,"5,000,000+",Free,0,Everyone,Art & Design,"August 1, 2018",1.2.4,4.0.3 and up,2018-08-01,2018,August,Art & Design
3,Sketch - Draw & Paint,ART_AND_DESIGN,4.5,215644,25M,"50,000,000+",Free,0,Teen,Art & Design,"June 8, 2018",Varies with device,4.2 and up,2018-06-08,2018,June,Art & Design
4,Pixel Draw - Number Art Coloring Book,ART_AND_DESIGN,4.3,967,2.8M,"100,000+",Free,0,Everyone,Art & Design;Creativity,"June 20, 2018",1.1,4.4 and up,2018-06-20,2018,June,Art & Design


In [ ]:
app.info()

<class 'pandas.core.frame.DataFrame'>
Index: 10358 entries, 0 to 10840
Data columns (total 17 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   App             10358 non-null  object        
 1   Category        10357 non-null  object        
 2   Rating          8893 non-null   float64       
 3   Reviews         10358 non-null  object        
 4   Size            10358 non-null  object        
 5   Installs        10358 non-null  object        
 6   Type            10357 non-null  object        
 7   Price           10358 non-null  object        
 8   Content Rating  10358 non-null  object        
 9   Genres          10357 non-null  object        
 10  Last Updated    10358 non-null  object        
 11  Current Ver     10350 non-null  object        
 12  Android Ver     10356 non-null  object        
 13  updated         10358 non-null  datetime64[ns]
 14  year            10358 non-null  int32         
 15  month  

In [ ]:
##converting reviews to number/int/float
app['rev_num'] = pd.to_numeric(app['Reviews'],errors='coerce')

In [ ]:
app.head()

,App,Category,Rating,Reviews,Size,Installs,Type,Price,Content Rating,Genres,Last Updated,Current Ver,Android Ver,updated,year,month,primary_genre,rev_num
0,Photo Editor & Candy Camera & Grid & ScrapBook,ART_AND_DESIGN,4.1,159,19M,"10,000+",Free,0,Everyone,Art & Design,"January 7, 2018",1.0.0,4.0.3 and up,2018-01-07,2018,January,Art & Design,159.0
1,Coloring book moana,ART_AND_DESIGN,3.9,967,14M,"500,000+",Free,0,Everyone,Art & Design;Pretend Play,"January 15, 2018",2.0.0,4.0.3 and up,2018-01-15,2018,January,Art & Design,967.0
2,"U Launcher Lite – FREE Live Cool Themes, Hide ...",ART_AND_DESIGN,4.7,87510,8.7M,"5,000,000+",Free,0,Everyone,Art & Design,"August 1, 2018",1.2.4,4.0.3 and up,2018-08-01,2018,August,Art & Design,87510.0
3,Sketch - Draw & Paint,ART_AND_DESIGN,4.5,215644,25M,"50,000,000+",Free,0,Teen,Art & Design,"June 8, 2018",Varies with device,4.2 and up,2018-06-08,2018,June,Art & Design,215644.0
4,Pixel Draw - Number Art Coloring Book,ART_AND_DESIGN,4.3,967,2.8M,"100,000+",Free,0,Everyone,Art & Design;Creativity,"June 20, 2018",1.1,4.4 and up,2018-06-20,2018,June,Art & Design,967.0


In [ ]:
app['Installs'].min(),app['Installs'].max()

('0', '500,000,000+')

In [ ]:
app[app['Installs']=='0']  ##only one app with zero installs


,App,Category,Rating,Reviews,Size,Installs,Type,Price,Content Rating,Genres,Last Updated,Current Ver,Android Ver,updated,year,month,primary_genre,rev_num
9148,Command & Conquer: Rivals,FAMILY,NaN,0,Varies with device,0,NaN,0,Everyone 10+,Strategy,"June 28, 2018",Varies with device,Varies with device,2018-06-28,2018,June,Strategy,0.0


In [ ]:
##converting app size to numeric
def siz(x):
  if 'M' in x:
    return float(x.replace('M',''))
  elif 'k' in x:
    return float(x.replace('k',''))
  else:
    return np.nan

In [ ]:
app['size_mb']=app['Size'].apply(siz)

In [ ]:
app.head()

,App,Category,Rating,Reviews,Size,Installs,Type,Price,Content Rating,Genres,Last Updated,Current Ver,Android Ver,updated,year,month,primary_genre,rev_num,size_mb
0,Photo Editor & Candy Camera & Grid & ScrapBook,ART_AND_DESIGN,4.1,159,19M,"10,000+",Free,0,Everyone,Art & Design,"January 7, 2018",1.0.0,4.0.3 and up,2018-01-07,2018,January,Art & Design,159.0,19.0
1,Coloring book moana,ART_AND_DESIGN,3.9,967,14M,"500,000+",Free,0,Everyone,Art & Design;Pretend Play,"January 15, 2018",2.0.0,4.0.3 and up,2018-01-15,2018,January,Art & Design,967.0,14.0
2,"U Launcher Lite – FREE Live Cool Themes, Hide ...",ART_AND_DESIGN,4.7,87510,8.7M,"5,000,000+",Free,0,Everyone,Art & Design,"August 1, 2018",1.2.4,4.0.3 and up,2018-08-01,2018,August,Art & Design,87510.0,8.7
3,Sketch - Draw & Paint,ART_AND_DESIGN,4.5,215644,25M,"50,000,000+",Free,0,Teen,Art & Design,"June 8, 2018",Varies with device,4.2 and up,2018-06-08,2018,June,Art & Design,215644.0,25.0
4,Pixel Draw - Number Art Coloring Book,ART_AND_DESIGN,4.3,967,2.8M,"100,000+",Free,0,Everyone,Art & Design;Creativity,"June 20, 2018",1.1,4.4 and up,2018-06-20,2018,June,Art & Design,967.0,2.8


In [ ]:
app['size_mb'].value_counts().head()

,count
size_mb,
11.0,189
13.0,186
12.0,186
14.0,183
15.0,174


In [ ]:
app['Price'].unique()

array(['0', '$4.99', '$3.99', '$6.99', '$1.49', '$2.99', '$7.99', '$5.99',
       '$3.49', '$1.99', '$9.99', '$7.49', '$0.99', '$9.00', '$5.49',
       '$10.00', '$24.99', '$11.99', '$79.99', '$16.99', '$14.99',
       '$1.00', '$29.99', '$12.99', '$2.49', '$10.99', '$1.50', '$19.99',
       '$15.99', '$33.99', '$74.99', '$39.99', '$3.95', '$4.49', '$1.70',
       '$8.99', '$2.00', '$3.88', '$25.99', '$399.99', '$17.99',
       '$400.00', '$3.02', '$1.76', '$4.84', '$4.77', '$1.61', '$2.50',
       '$1.59', '$6.49', '$1.29', '$5.00', '$13.99', '$299.99', '$379.99',
       '$37.99', '$18.99', '$389.99', '$19.90', '$8.49', '$1.75',
       '$14.00', '$4.85', '$46.99', '$109.99', '$154.99', '$3.08',
       '$2.59', '$4.80', '$1.96', '$19.40', '$3.90', '$4.59', '$15.46',
       '$3.04', '$4.29', '$2.60', '$3.28', '$4.60', '$28.99', '$2.95',
       '$2.90', '$1.97', '$200.00', '$89.99', '$2.56', '$30.99', '$3.61',
       '$394.99', '$1.26', '0.0', '$1.20', '$1.04'], dtype=object)

In [ ]:
def pric(x):
  if '$' in x:
    return float(x.replace('$',''))
  else:
    return float(x)

In [ ]:
##converting price to numeric
app['price_dol'] = app['Price'].apply(pric)

In [ ]:
app['price_dol'].unique()

array([  0.  ,   4.99,   3.99,   6.99,   1.49,   2.99,   7.99,   5.99,
         3.49,   1.99,   9.99,   7.49,   0.99,   9.  ,   5.49,  10.  ,
        24.99,  11.99,  79.99,  16.99,  14.99,   1.  ,  29.99,  12.99,
         2.49,  10.99,   1.5 ,  19.99,  15.99,  33.99,  74.99,  39.99,
         3.95,   4.49,   1.7 ,   8.99,   2.  ,   3.88,  25.99, 399.99,
        17.99, 400.  ,   3.02,   1.76,   4.84,   4.77,   1.61,   2.5 ,
         1.59,   6.49,   1.29,   5.  ,  13.99, 299.99, 379.99,  37.99,
        18.99, 389.99,  19.9 ,   8.49,   1.75,  14.  ,   4.85,  46.99,
       109.99, 154.99,   3.08,   2.59,   4.8 ,   1.96,  19.4 ,   3.9 ,
         4.59,  15.46,   3.04,   4.29,   2.6 ,   3.28,   4.6 ,  28.99,
         2.95,   2.9 ,   1.97, 200.  ,  89.99,   2.56,  30.99,   3.61,
       394.99,   1.26,   1.2 ,   1.04])

### Data Wrangling Code for User review data

In [ ]:
rev.duplicated().sum()

np.int64(33616)

In [ ]:
rev.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 64295 entries, 0 to 64294
Data columns (total 5 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   App                     64295 non-null  object 
 1   Translated_Review       37427 non-null  object 
 2   Sentiment               37432 non-null  object 
 3   Sentiment_Polarity      37432 non-null  float64
 4   Sentiment_Subjectivity  37432 non-null  float64
dtypes: float64(2), object(3)
memory usage: 2.5+ MB


In [ ]:
rev[rev.duplicated()].head()

,App,Translated_Review,Sentiment,Sentiment_Polarity,Sentiment_Subjectivity
7,10 Best Foods for You,NaN,NaN,NaN,NaN
15,10 Best Foods for You,NaN,NaN,NaN,NaN
78,10 Best Foods for You,Good,Positive,0.7,0.600000
79,10 Best Foods for You,Good,Positive,0.7,0.600000
100,10 Best Foods for You,I like eat delicious food. That's I'm cooking ...,Positive,1.0,0.533333


In [ ]:
##dropping dupicates
rev.drop_duplicates(keep='first',inplace=True)


In [ ]:
rev.info() ##checking for null values

<class 'pandas.core.frame.DataFrame'>
Index: 30679 entries, 0 to 64236
Data columns (total 5 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   App                     30679 non-null  object 
 1   Translated_Review       29692 non-null  object 
 2   Sentiment               29697 non-null  object 
 3   Sentiment_Polarity      29697 non-null  float64
 4   Sentiment_Subjectivity  29697 non-null  float64
dtypes: float64(2), object(3)
memory usage: 1.4+ MB


In [ ]:
rev.dropna(inplace=True)  ##dropping null values

In [ ]:
rev.info()

<class 'pandas.core.frame.DataFrame'>
Index: 29692 entries, 0 to 64230
Data columns (total 5 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   App                     29692 non-null  object 
 1   Translated_Review       29692 non-null  object 
 2   Sentiment               29692 non-null  object 
 3   Sentiment_Polarity      29692 non-null  float64
 4   Sentiment_Subjectivity  29692 non-null  float64
dtypes: float64(2), object(3)
memory usage: 1.4+ MB


**Variable engineering**

In [ ]:
rev.head()

,App,Translated_Review,Sentiment,Sentiment_Polarity,Sentiment_Subjectivity
0,10 Best Foods for You,I like eat delicious food. That's I'm cooking ...,Positive,1.00,0.533333
1,10 Best Foods for You,This help eating healthy exercise regular basis,Positive,0.25,0.288462
3,10 Best Foods for You,Works great especially going grocery store,Positive,0.40,0.875000
4,10 Best Foods for You,Best idea us,Positive,1.00,0.300000
5,10 Best Foods for You,Best way,Positive,1.00,0.300000


In [ ]:
rev['review_len'] = rev['Translated_Review'].str.len()


In [ ]:
rev.head()

,App,Translated_Review,Sentiment,Sentiment_Polarity,Sentiment_Subjectivity,review_len
0,10 Best Foods for You,I like eat delicious food. That's I'm cooking ...,Positive,1.00,0.533333,122
1,10 Best Foods for You,This help eating healthy exercise regular basis,Positive,0.25,0.288462,47
3,10 Best Foods for You,Works great especially going grocery store,Positive,0.40,0.875000,42
4,10 Best Foods for You,Best idea us,Positive,1.00,0.300000,12
5,10 Best Foods for You,Best way,Positive,1.00,0.300000,8


In [ ]:
##saving both files
app.to_csv('cw app data.csv',index=False)
rev.to_csv('cw user_rev data.csv',index=False)

In [ ]:
files.download('cw app data.csv')
files.download('cw user_rev data.csv')

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

### What all manipulations have you done and insights you found?

1. Playstore app data and user review data has null values and duplicate values, these were removed/droped --> **DATA CLEANING**.
2. creation of new columns like updated,price_dol,primary_gen etc in play store data and columns like review_len in user review data is-->**VARIABLE ENGINEERING**.
3. Other manipulations will be done during visualization.

## ***4. Data Vizualization, Storytelling & Experimenting with charts : Understand the relationships between variables***

#### Chart - 1 UNIVARIATE ANALYSIS

In [ ]:
app=pd.read_csv('/content/drive/MyDrive/PLAYSTORE/cw app data (1).csv')
rev = pd.read_csv('/content/drive/MyDrive/PLAYSTORE/cw user_rev data (1).csv')

In [ ]:
del app['Unnamed: 0']

In [ ]:
app.head()

,App,Category,Rating,Reviews,Size,Installs,Type,Price,Content Rating,Genres,Last Updated,Current Ver,Android Ver,updated,year,month,primary_genre,rev_num,install_num,size_mb,price_dol
0,Photo Editor & Candy Camera & Grid & ScrapBook,ART_AND_DESIGN,4.1,159.0,19M,"10,000+",Free,0,Everyone,Art & Design,"January 7, 2018",1.0.0,4.0.3 and up,2018-01-07,2018,January,Art & Design,159.0,"10,000",19.0,0.0
1,Coloring book moana,ART_AND_DESIGN,3.9,967.0,14M,"500,000+",Free,0,Everyone,Art & Design;Pretend Play,"January 15, 2018",2.0.0,4.0.3 and up,2018-01-15,2018,January,Art & Design,967.0,"500,000",14.0,0.0
2,"U Launcher Lite – FREE Live Cool Themes, Hide ...",ART_AND_DESIGN,4.7,87510.0,8.7M,"5,000,000+",Free,0,Everyone,Art & Design,"August 1, 2018",1.2.4,4.0.3 and up,2018-08-01,2018,August,Art & Design,87510.0,"5,000,000",8.7,0.0
3,Sketch - Draw & Paint,ART_AND_DESIGN,4.5,215644.0,25M,"50,000,000+",Free,0,Teen,Art & Design,"June 8, 2018",Varies with device,4.2 and up,2018-06-08,2018,June,Art & Design,215644.0,"50,000,000",25.0,0.0
4,Pixel Draw - Number Art Coloring Book,ART_AND_DESIGN,4.3,967.0,2.8M,"100,000+",Free,0,Everyone,Art & Design;Creativity,"June 20, 2018",1.1,4.4 and up,2018-06-20,2018,June,Art & Design,967.0,"100,000",2.8,0.0


In [ ]:
data=app['Category'].value_counts().reset_index()

In [ ]:
data.head()

,Category,count
0,FAMILY,1943
1,GAME,1121
2,TOOLS,843
3,BUSINESS,427
4,MEDICAL,408


In [ ]:
## Barplot for app category
fig = px.bar(
    data,
    x='Category',
    y='count',
    color='Category',
    text_auto=True,
    title='Category-wise Distribution/count'
)

fig.update_layout(
    showlegend=False,
    title_font=dict(size=20,weight='bold'),
    font=dict(size=12,weight='bold'),
    xaxis_title='Categories',
    yaxis_title='Count',
    bargap=0.2
)

fig.update_traces(
    textposition='outside',
    marker_line_width=1.2
)

fig.show()


##### 1. Why did you pick the specific chart?

Here **Barchart** is used to **compare the app categories with respect to their count values.**

##### 2. What is/are the insight(s) found from the chart?

1. **Family app category** has the **highest count value** of **1943**,it is followed by **Game(1121)** and **Tools(843)**. Suggesting that these are the **most popular/used app categories** wrt users.

2. There is a **significant/sharp drop in count values** after the above mentioned catgories. **Business,medical and productivity** have **moderate** count value **above 400**.

3. The **lower count** values are for categories like **Beauty(53),Parenting,comics(60),Events(60) and Art and design(65)** suggesitng these app categories are **least prefered** by the users **or** they were **developed recently**.

##### 3. Will the gained insights help creating a positive business impact?
Are there any insights that lead to negative growth? Justify with specific reason.

**POSITIVES:**
1. **Huge market/demand** for the app catgeories with high count, **Niche app categories** have **low competition**, allows the new apps to gain visibility.
2. **Diverse app category** presence on the platform. **Family,education,medcial,Health and fitness app categories drive repeat and long-term usage.**

**NEGATIVES:**
1. **Huge competition** in **top app categories**,can lead to **Market stauration**. **Recognition** for the **newly made apps** is **difficult** and **slow growth**. User acquisation can be **costly** in these categories.

2. There can be some **Scalability issues** in the **low-count categories**,slows growth and revenue **despite low competition**.

#### Chart - 2 Bivariate analysis

In [ ]:
data = app.groupby('Category').agg({'Rating':'mean'}).reset_index()

In [ ]:
data.sort_values(by='Rating',ascending=False,ignore_index=True,inplace=True)

In [ ]:
data.head()

,Category,Rating
0,EVENTS,4.435556
1,EDUCATION,4.375969
2,ART_AND_DESIGN,4.358065
3,BOOKS_AND_REFERENCE,4.347458
4,PERSONALIZATION,4.333871


In [ ]:
# Chart - 2 visualization code
fig = px.bar(
    data,
    x='Category',
    y='Rating',
    color='Category',
    text_auto=True,
    title='Category-wise Distribution of Ratings'
)

fig.update_layout(
    showlegend=False,
    title_font_size=20,
    xaxis_title='Category',
    yaxis_title='Avg.Ratings',
    title_font=dict(size=20,weight='bold'),
    font=dict(size=12,weight='bold'),
    bargap=0.25
)

fig.update_traces(
    textposition='outside',
    marker_line_width=1.2
)

fig.show()

##### 1. Why did you pick the specific chart?

**Bar chart** to **compare apps/app catgories wrt Ratings.**

##### 2. What is/are the insight(s) found from the chart?

1. **Majority** of app categories have **avg.rating above 4.0**, **Dating apps** have avg.rating of **3.97**.
2. **Highest** amongst the ratings (above 4.3) are **Events,Education, Art & design, Books & reference, Personelization and Parenting** apps. Suggesting users favour information and content driven apps.
3. There is **no significant drop** in avg.ratings accross categories. Therefore ratings alone can not differentiate categories.

##### 3. Will the gained insights help creating a positive business impact?
Are there any insights that lead to negative growth? Justify with specific reason.

**POSITIVES**:
1. **Higher ratings** for apps on the paltform suggests **Playsotre is a reliabe,mature and trustworthy.** These app categories improve **user satisfaction, allow premium pricing and special offer reccomendations to users**

2. **Subscription acquisation** through **highly rated apps** like(Eductaion,Books and Art). **Engagement** through high utitlity apps like **Maps & Navigation,communication and Tools**

#### Chart - 3 Bivariate analysis

In [ ]:
#data preperation
data = app.groupby('Installs')['Type'].value_counts().reset_index()

In [ ]:
data.head()

,Installs,Type,count
0,0+,Paid,10
1,0+,Free,4
2,1+,Free,46
3,1+,Paid,21
4,"1,000+",Free,752


In [ ]:
data.sort_values(by='count',ascending=False,inplace=True,ignore_index=True)

In [ ]:
# Chart - 3 visualization code
fig = px.bar(
    data,
    x='Installs',
    y='count',
    color='Type',
    title='Bar plot for Installs wrt type'
)

fig.update_layout(
    title_font_size=20,
    title_font_weight = 'bold',
    font= dict(size=12,weight='bold'),
    yaxis_title='count of installs '
)

fig.show()


##### 1. Why did you pick the specific chart?

**Barchart** to compare **installment buckets wrt type** (Free or paid apps).

##### 2. What is/are the insight(s) found from the chart?

1. **Free apps dominate all install buckets**, suggesting the users prefer **zero-cost entry**. **Paid apps are very low** in all install buckets may be **due to upfront pricing**.
2. Apps with **1M installs are more (1467)**,followed by **10M(1129)** and **100K(1004) install apps**.
3. Vry **few Apps with 1B and 500M installs(high install bucket)**.These high install buckets is achevied by **free apps**.
4. In the **mid-range installs** bucket **paid apps** show **limited** presence.

##### 3. Will the gained insights help creating a positive business impact?
Are there any insights that lead to negative growth? Justify with specific reason.

 **POSITIVES:**
 1. **Rapid user acquisation** through **free apps**, making them ideal for **brand bulding,marketing, and promotion.**
 2. **Paid apps** for **users** with **strong commitment** and **high engagement**.

3. **Volume driven apps--> Free type.**
4. **Value driven apps --> Paid type.**

5. **HYBRID MODELS(free + in-app purchases)** for **scale/volume and value/revenue.** eg: Canvas,inshot etc

**NEGATIVES:**
1. **Paid apps face lower discoverability**,scalability issues and reduced user expereince.
2. **Free apps** must depend on **adds for revenue**, which will lead** user dissapointment** and reduce installs or **increase bad reviews**.

#### Chart - 4 Bivariate analysis

In [ ]:
app.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10358 entries, 0 to 10357
Data columns (total 21 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   App             10358 non-null  object 
 1   Category        10357 non-null  object 
 2   Rating          8893 non-null   float64
 3   Reviews         10358 non-null  float64
 4   Size            10358 non-null  object 
 5   Installs        10358 non-null  object 
 6   Type            10357 non-null  object 
 7   Price           10358 non-null  object 
 8   Content Rating  10358 non-null  object 
 9   Genres          10357 non-null  object 
 10  Last Updated    10358 non-null  object 
 11  Current Ver     10350 non-null  object 
 12  Android Ver     10356 non-null  object 
 13  updated         10358 non-null  object 
 14  year            10358 non-null  int64  
 15  month           10358 non-null  object 
 16  primary_genre   10357 non-null  object 
 17  rev_num         10358 non-null 

In [ ]:
app['size_mb'].fillna('Varies with device',inplace=True)

/tmp/ipython-input-1360191940.py:1: FutureWarning:

A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.



/tmp/ipython-input-1360191940.py:1: FutureWarning:

Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value 'Varies with device' has dtype incompatible with float64, please explicitly cast to a compatible dtype first.



In [ ]:
app.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10358 entries, 0 to 10357
Data columns (total 21 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   App             10358 non-null  object 
 1   Category        10357 non-null  object 
 2   Rating          8893 non-null   float64
 3   Reviews         10358 non-null  float64
 4   Size            10358 non-null  object 
 5   Installs        10358 non-null  object 
 6   Type            10357 non-null  object 
 7   Price           10358 non-null  object 
 8   Content Rating  10358 non-null  object 
 9   Genres          10357 non-null  object 
 10  Last Updated    10358 non-null  object 
 11  Current Ver     10350 non-null  object 
 12  Android Ver     10356 non-null  object 
 13  updated         10358 non-null  object 
 14  year            10358 non-null  int64  
 15  month           10358 non-null  object 
 16  primary_genre   10357 non-null  object 
 17  rev_num         10358 non-null 

In [ ]:
data = app.groupby('year')['rev_num'].sum().reset_index()

In [ ]:
data.head()

,year,rev_num
0,2010,1415.0
1,2011,235800.0
2,2012,270362.0
3,2013,3573548.0
4,2014,6571229.0


In [ ]:
data1 = data[data['year']<=2016]
data2 = data[data['year']>2016]

In [ ]:
# Chart - 4 visualization code

fig = px.line(
    data1,
    x='year',
    y='rev_num',
    title='Line plot for reviews vs year')

fig.update_layout(
    title_font_size=20,
    title_font_weight = 'bold',
    font= dict(size=12,weight='bold'),
    yaxis_title='no.of reviews'
)

fig.show()

fig = px.line(
    data2,
    x='year',
    y='rev_num',
    title='Line plot for reviews vs year')

fig.update_layout(
    title_font_size=20,
    title_font_weight = 'bold',
    font= dict(size=12,weight='bold'),
    yaxis_title='no.of reviews'
)

fig.show()


##### 1. Why did you pick the specific chart?

**Line chart is used to visualize trends over time**. Here in this we are going to visualize how review count changed over time.

##### 2. What is/are the insight(s) found from the chart?

1. There is a **steady increase in no.of reviews through out the years**. In the year **2010**, the reviews are around **1415** and **till 2012** the no.of reviews are **below 250k**. After **2012** the no.of reviews **enters into millions**/exponential growth, till **2017** with **103 million reviews**.

2. After **2017** the no.of reviews enters the **Billions club** which is a **huge jump,** suggesting the **huge increase** in the **no.of android phones** by the users and **their affordability**,**stable internet/wifi connections accross the world** and **more user engagement** on the platform/playstore.

##### 3. Will the gained insights help creating a positive business impact?
Are there any insights that lead to negative growth? Justify with specific reason.

**POSITIVES:**
1. **Strong user engagement** post 2017, users **actively intercting** on the platform, **increasing trust** and **credibility of the platform**.

2. **Faster app improvement/upgrades** by app developers with the help of **increasing no.of reviews and feedbacks.**

3. **Higher monetization potential** due to **higher engagement,in app purchases** and **ads**.

**NEGATIVES**:
* Increase in **Fake,abusive and meaningless reviews leads to noise,redcued review filtering effeciency.**


#### Chart - 5 Bivariate analysis

In [ ]:
data = app.groupby('Category')['rev_num'].sum().reset_index()

In [ ]:
data.sort_values(by='rev_num',ascending=False,inplace=True,ignore_index=True)

In [ ]:
# Chart - 5 visualization code
fig = px.bar(
    data,
    x='Category',
    y='rev_num',
    color='Category',
    title='Bar plot for App category vs No.of Reviews'
)

fig.update_layout(
    title_font_size=20,
    title_font_weight = 'bold',
    font= dict(size=12,weight='bold'),
    yaxis_title='No.of Reviews'
)

fig.show()

##### 1. Why did you pick the specific chart?

**Barchart is used to compare categories wrt no.of reviews.**


##### 2. What is/are the insight(s) found from the chart?

1. **Gaming apps** have **highest no.of reviews(1.4B)** suggesting higher user engagement. This is followed by **communication apps(600+M),social apps(500+M) and family apps(390+M)** are moderate engagement apps. There is a significant/sharp drop in no.of reviews when compared with gaming apps.

2. App categories like **Beauty,Events,parenting,medical etc have very low no.of reviews (1-5M) suggesting less user engagement/lesser installs.**


##### 3. Will the gained insights help creating a positive business impact?
Are there any insights that lead to negative growth? Justify with specific reason.

**POSITIVES**:
1. **High revenue potential** app categories like Games,communication,social and family through **ads,in app purchases and subscriptions.**
2. **Focus on ranking,reccomendations and promotions** based on high-impact categories.
3. **Review rich**/Data rich categories can be used for **analysis,prediction and personalization.**

**NEGATIVES:**
1. **High-review categories overshadow the low-review categories** in discoverability,scale,engagement and revenue. **Limited feedback** on low-review categorie, which leads to **limited efforst** in improving the** user satisfaction,app upgrade etc.**

#### Chart - 6 Bivariate analysis

In [ ]:
data = app[app['price_dol']>0]

In [ ]:
data.head()

,App,Category,Rating,Reviews,Size,Installs,Type,Price,Content Rating,Genres,Last Updated,Current Ver,Android Ver,updated,year,month,primary_genre,rev_num,install_num,size_mb,price_dol
233,TurboScan: scan documents and receipts in PDF,BUSINESS,4.7,11442.0,6.8M,"100,000+",Paid,$4.99,Everyone,Business,"March 25, 2018",1.5.2,4.0 and up,2018-03-25,2018,March,Business,11442.0,"100,000",6.8,4.99
234,Tiny Scanner Pro: PDF Doc Scan,BUSINESS,4.8,10295.0,39M,"100,000+",Paid,$4.99,Everyone,Business,"April 11, 2017",3.4.6,3.0 and up,2017-04-11,2017,April,Business,10295.0,"100,000",39.0,4.99
390,Puffin Browser Pro,COMMUNICATION,4.0,18247.0,Varies with device,"100,000+",Paid,$3.99,Everyone,Communication,"July 5, 2018",7.5.3.20547,4.1 and up,2018-07-05,2018,July,Communication,18247.0,"100,000",Varies with device,3.99
429,"Moco+ - Chat, Meet People",DATING,4.2,1545.0,Varies with device,"10,000+",Paid,$3.99,Mature 17+,Dating,"June 19, 2018",2.6.139,4.1 and up,2018-06-19,2018,June,Dating,1545.0,"10,000",Varies with device,3.99
430,Calculator,DATING,2.6,57.0,6.2M,"1,000+",Paid,$6.99,Everyone,Dating,"October 25, 2017",1.1.6,4.0 and up,2017-10-25,2017,October,Dating,57.0,"1,000",6.2,6.99


In [ ]:
##chart-6 visualization code
fig = px.box(
    data,
    x='Category',
    y='price_dol',
    color='Category',
    title='Box plot for Categories vs price in dollars'
)

fig.update_layout(
    title_font_size=20,
    title_font_weight = 'bold',
    font= dict(size=12,weight='bold'),
    yaxis_title='Price in dollars'
)

fig.show()


##### 1. Why did you pick the specific chart?

**Box plot is used to visualize the distribution and identify outliers**. Here in this boxplot we are going to **visualize the distribution of paid app categories wrt their prices in dollars**.

##### 2. What is/are the insight(s) found from the chart?

1. Highest/**Most spread distribution of price** is observed for **Lifestyle and Finance**. **Lifestyle** apps has a **median of 4.99 dollars** and **highest price of 400 dollars**. **Finance** apps has a **median price of 28.99** and **highest price of 399.99 dollars**. Compare to these two top app categories **all other app categories have lower median prices almost close to zero** with very few/no outliers.

2. **Games,Family,Photography and medical** apps have **extreme price outliers**(some apps have **price close to 400$**)

##### 3. Will the gained insights help creating a positive business impact?
Are there any insights that lead to negative growth? Justify with specific reason.

**POSITIVES:**
1. **Lower pricing allows --> low entry barriers for the users--> high installs,better market penetration**

2. **Premium pricing** for app categories like **Medical, Finance and Education**.

**NEGATIVES**:
1. **Scalability,Discoverabilty** of the **paid apps** is very difficult to acheive. **App underpricing** to remain in the competition.

#### Chart - 7 Bivariate analysis

In [ ]:
data = app[app['price_dol']>0]

In [ ]:
data = data.groupby('Installs')['price_dol'].mean().reset_index()

In [ ]:
data.sort_values(by='price_dol',ascending = False,ignore_index=True,inplace=True)

In [ ]:
data.head()

,Installs,price_dol
0,0+,120.541000
1,"5,000+",21.799692
2,"10,000+",20.155932
3,"1,000+",17.429209
4,"50,000+",14.338293


In [ ]:
# Chart - 7 visualization code
fig = px.bar(
    data,
    x='Installs',
    y='price_dol',
    color='Installs',
    title='Bar plot for Installs vs price in dollars'
)

fig.update_layout(
    title_font_size=20,
    title_font_weight = 'bold',
    font= dict(size=12,weight='bold'),
    yaxis_title='avg.price in dollars '
)

fig.show()

##### 1. Why did you pick the specific chart?

**Bar chart to visusalze the price wrt to installs buckets, which can be used as revenue proxy.**

##### 2. What is/are the insight(s) found from the chart?

1. Apps with **zero installs have the highest average pricing**,suggesting **very expensive/high price** entry apps have almost **zero installs/zero users**.

2. Apps with **high no.of installs(1M,10M,0.5M) have low pricing** suggesting high no.of users and high user enagagement. Apps with **1B OR 500M installs are free**(Not shown here).

3. Apps with **Install buckets 5k,10k,50k and 100k** have ***avg.pricings*** of **21.79, 20.15, 14.33 and 9.58 dollars** respectively.(moderate to low pricing)

4. **Pricing is negatively related to scaling and installs**.


##### 3. Will the gained insights help creating a positive business impact?
Are there any insights that lead to negative growth? Justify with specific reason.

POSITIVES:
1. **Max reach --> Low pricing**, **clear pricing for app developers** to adjust price strategies (in-app purchases) based on ths data, wrt their goals.

2. **Specialised apps--> High-prices--> Small-targeted users**.

NEGATIVES:
* **Free apps have revenue pressure**,they have to depend on in app purchases and ads (high number).

#### Chart - 8 Bivariate analysis

In [ ]:
# Chart - 8 visualization code
# Chart - 4 scatter plot
fig = px.scatter(
    app,
    y='size_mb',
    x='Rating',
    color='Type',
    size_max=50,
    labels={
        'size_mb': 'Size in MB',
        'Rating':'App ratings'

    },
    title='App size vs app-ratings'
)

fig.update_traces(
    marker=dict(
        opacity=0.75,
        line=dict(width=1, color='DarkSlateGrey')
    )
)

fig.update_layout(
    title_font_size=20,
    title_font_weight='bold',
    xaxis=dict(
        showgrid=True,
        gridcolor='lightgrey'
    ),
    yaxis=dict(
        showgrid=True,
        gridcolor='lightgrey'
    ),
    legend_title_text='Region',
    height=500,
    template='plotly_white'
)

fig.show()


##### 1. Why did you pick the specific chart?

**Scatterplo**t to visualize the relationship between two continous variables. Here we are exploring the **realtionship between App size and app ratings.**

##### 2. What is/are the insight(s) found from the chart?

1. **Most Paid apps have a size of below 120 mb and ratings above 3.0**. **Most of the apps** (free or paid) have **ratings above 3.5**.

2. Compared to the free apps, **very few paid apps are big in size**(700-1000 Mb). **Free apps dominate both size and ratings**
3. There is no clear correlation between appsize and app-ratings,(apps with range of sizes have range of ratings). Highly rated apps have various sizes, suggesting **users prefer functionality** and content over size.
4. **Wider rating dispersion for large apps**.

##### 3. Will the gained insights help creating a positive business impact?
Are there any insights that lead to negative growth? Justify with specific reason.

POSITIVES:
1. **No app size constrains** for developers, **higher ratings** can be acheived by delivering quality/**rich functionality** especially for **free-feature rich apps**.

2. **Paid apps mostly benefit from being lightweight**(low size).

NEGATIVES:
1. **Paid apps and large size** --> more resources,testing and updates --> more expectations in functionality.

#### Chart - 9 MULTIVARIATE ANALYSIS


In [ ]:
data1 = app.groupby('year')['App'].value_counts().reset_index()
data2 = app.groupby('year')['Rating'].mean().reset_index()

In [ ]:
data2.head()

,year,Rating
0,2010,4.200000
1,2011,3.966667
2,2012,3.785714
3,2013,4.069565
4,2014,4.036757


In [ ]:
data1.head()

,year,App,count
0,2010,FML F*ck my life + widget,1
1,2011,AH Alarm Panel,1
2,2011,Apk Installer,1
3,2011,Band O'Clock,1
4,2011,Bubble,1


In [ ]:
data1= data1.groupby('year')['count'].sum().reset_index()

In [ ]:
data1.head()

,year,count
0,2010,1
1,2011,15
2,2012,26
3,2013,108
4,2014,204


In [ ]:
app.head()

,App,Category,Rating,Reviews,Size,Installs,Type,Price,Content Rating,Genres,Last Updated,Current Ver,Android Ver,updated,year,month,primary_genre,rev_num,install_num,size_mb,price_dol
0,Photo Editor & Candy Camera & Grid & ScrapBook,ART_AND_DESIGN,4.1,159.0,19M,"10,000+",Free,0,Everyone,Art & Design,"January 7, 2018",1.0.0,4.0.3 and up,2018-01-07,2018,January,Art & Design,159.0,"10,000",19.0,0.0
1,Coloring book moana,ART_AND_DESIGN,3.9,967.0,14M,"500,000+",Free,0,Everyone,Art & Design;Pretend Play,"January 15, 2018",2.0.0,4.0.3 and up,2018-01-15,2018,January,Art & Design,967.0,"500,000",14.0,0.0
2,"U Launcher Lite – FREE Live Cool Themes, Hide ...",ART_AND_DESIGN,4.7,87510.0,8.7M,"5,000,000+",Free,0,Everyone,Art & Design,"August 1, 2018",1.2.4,4.0.3 and up,2018-08-01,2018,August,Art & Design,87510.0,"5,000,000",8.7,0.0
3,Sketch - Draw & Paint,ART_AND_DESIGN,4.5,215644.0,25M,"50,000,000+",Free,0,Teen,Art & Design,"June 8, 2018",Varies with device,4.2 and up,2018-06-08,2018,June,Art & Design,215644.0,"50,000,000",25.0,0.0
4,Pixel Draw - Number Art Coloring Book,ART_AND_DESIGN,4.3,967.0,2.8M,"100,000+",Free,0,Everyone,Art & Design;Creativity,"June 20, 2018",1.1,4.4 and up,2018-06-20,2018,June,Art & Design,967.0,"100,000",2.8,0.0


In [ ]:
data3 = app.groupby('Content Rating')['App'].value_counts().reset_index()

In [ ]:
data3 = data3.groupby('Content Rating')['count'].sum().reset_index()

In [ ]:
data3.rename(columns={'count':'no.of apps'},inplace=True)

In [ ]:
data3.head()

,Content Rating,no.of apps
0,Unrated,2
1,Adults only 18+,3
2,Everyone 10+,377
3,Mature 17+,447
4,Teen,1146


In [ ]:
data3.sort_values(by='no.of apps',ascending=True,ignore_index=True,inplace=True)

In [ ]:
# Chart - 9 visualization code
# Chart - 8 visualization code
import plotly.graph_objects as go
from plotly.subplots import make_subplots

fig = make_subplots(
    rows=1, cols=3,
    subplot_titles=(
        'No.of Apps wrt Year/Time',
        'Rating vs Year/Time',
        'content rating vs no.of apps ',
    )
)

# Line
fig.add_trace(
    go.Scatter(x=data1['year'],y=data1['count'],name='counts'),
    row=1, col=1)

# line
fig.add_trace(
    go.Scatter(x=data2['year'],y=data2['Rating'], name='ratings'),
    row=1, col=2)

# barplot
fig.add_trace(
     go.Bar(x=data3['Content Rating'],y=data3['no.of apps'],name='no.of apps'),
    row=1,col=3)

fig.update_layout(
    title='Subplots',
    title_font_size=20,
    title_font_weight = 'bold',
    font= dict(size=12,weight='bold')
)

fig.show()



##### 1. Why did you pick the specific chart?

  **Subplots to visualize multiple charts at a time for analysis**. Here in this chart we have used two line plot (for trends over time) and one bar graph(for category comparision)

##### 2. What is/are the insight(s) found from the chart?

1. In **2010** there is only **one app** on the playstore. **After 2010 there is a steady increase in the no.of apps** through out the years and **grew exponentially after 2014,** and reached a peak count of **6935 apps in 2018.** This indicates massive expansions,more android phones,stable internet and more users.

2. After 2010 the **ratings have dropped till 2012** (from 4.2 to 3.78). After this, **there is a significant increase in the avg.ratings, reaching 4.245 in 2018** (with some lows and highs). There **isn't heavy deviation in ratings** (most ratings between 3.8-4.2) indicating **good app quality** on the platform.
3. Content Rating **'Everyone' dominates the platforms ecosystem** with **8393 apps**. other categories have very low no.of apps **Teen(1146 apps)** and **mature 17+(447 apps)** and there minimal adult only apps.

##### 3. Will the gained insights help creating a positive business impact?
Are there any insights that lead to negative growth? Justify with specific reason.

**POSITIVES:**
1. **Rapid app growth on the platform --> no barriers for devalopers --> better scalability**.

2. **No deterioration** in the **app quality**. **Broad user/audience** reach with Apps content rating as **'Everyone'**.

**NEGATIVES:**
1. **Limited diversity** in content rating. **Overreliance on Everyone-rated apps** may **limit innovations** and development of **specialised apps like Dating**,Communication,social-media apps etc.

#### Chart - 10 Multivariable analysis using heatmap


In [ ]:
app.head()

,App,Category,Rating,Reviews,Size,Installs,Type,Price,Content Rating,Genres,Last Updated,Current Ver,Android Ver,updated,year,month,primary_genre,rev_num,install_num,size_mb,price_dol
0,Photo Editor & Candy Camera & Grid & ScrapBook,ART_AND_DESIGN,4.1,159.0,19M,"10,000+",Free,0,Everyone,Art & Design,"January 7, 2018",1.0.0,4.0.3 and up,2018-01-07,2018,January,Art & Design,159.0,"10,000",19.0,0.0
1,Coloring book moana,ART_AND_DESIGN,3.9,967.0,14M,"500,000+",Free,0,Everyone,Art & Design;Pretend Play,"January 15, 2018",2.0.0,4.0.3 and up,2018-01-15,2018,January,Art & Design,967.0,"500,000",14.0,0.0
2,"U Launcher Lite – FREE Live Cool Themes, Hide ...",ART_AND_DESIGN,4.7,87510.0,8.7M,"5,000,000+",Free,0,Everyone,Art & Design,"August 1, 2018",1.2.4,4.0.3 and up,2018-08-01,2018,August,Art & Design,87510.0,"5,000,000",8.7,0.0
3,Sketch - Draw & Paint,ART_AND_DESIGN,4.5,215644.0,25M,"50,000,000+",Free,0,Teen,Art & Design,"June 8, 2018",Varies with device,4.2 and up,2018-06-08,2018,June,Art & Design,215644.0,"50,000,000",25.0,0.0
4,Pixel Draw - Number Art Coloring Book,ART_AND_DESIGN,4.3,967.0,2.8M,"100,000+",Free,0,Everyone,Art & Design;Creativity,"June 20, 2018",1.1,4.4 and up,2018-06-20,2018,June,Art & Design,967.0,"100,000",2.8,0.0


In [ ]:
data = app[~(app['size_mb']=='Varies with device')]

In [ ]:

data = data[['rev_num','size_mb','price_dol','Rating']].corr()

In [ ]:
# Chart - 10 visualization code
# Correlation Heatmap visualization code
fig = px.imshow(
    data,
    text_auto='.2f',
    color_continuous_scale='cividis',
    title='Correlation Heatmap',
    zmin=-1,
    zmax=1,
    aspect='auto'
)
fig.update_traces(
    textfont=dict(
        size=13,
        color='black'
    ))

fig.update_layout(
    font=dict(size=22, family='Arial Black'),
    width=1200,
    height=600,
    margin=dict(l=120, r=80, t=100, b=80)
)

fig.show()

##### 1. Why did you pick the specific chart?

**Correlation heatmap to find out the strength and direction of correlation between multiple varibale at a time.**


##### 2. What is/are the insight(s) found from the chart?

There is **no correlation/very minimal(close to zero)** among the variables in the heatmap. All variables are independent of others in the chart.

##### 3. Will the gained insights help creating a positive business impact?
Are there any insights that lead to negative growth? Justify with specific reason.

Answer Here

#### Chart - 11 Univariate analysis

In [ ]:
rev.head()


,Unnamed: 0,App,Translated_Review,Sentiment,Sentiment_Polarity,Sentiment_Subjectivity,review_len
0,0,10 Best Foods for You,I like eat delicious food. That's I'm cooking ...,Positive,1.00,0.533333,122
1,1,10 Best Foods for You,This help eating healthy exercise regular basis,Positive,0.25,0.288462,47
2,3,10 Best Foods for You,Works great especially going grocery store,Positive,0.40,0.875000,42
3,4,10 Best Foods for You,Best idea us,Positive,1.00,0.300000,12
4,5,10 Best Foods for You,Best way,Positive,1.00,0.300000,8


In [ ]:
del rev['Unnamed: 0']

In [ ]:
rev['Sentiment'].value_counts()

,count
Sentiment,
Positive,19015
Negative,6321
Neutral,4356


In [ ]:
data = rev['Sentiment'].value_counts().reset_index()

In [ ]:
data.head()

,Sentiment,count
0,Positive,19015
1,Negative,6321
2,Neutral,4356


In [ ]:
# Chart - 11 visualization code
fig = px.pie(data,names='Sentiment',values='count',title='Pie chart for Review Sentiment proportion',hole=0.2)
fig.update_layout(
    title_font=dict(size=20,weight='bold'),
    font =dict(size=12,weight='bold')
)
fig.show()


##### 1. Why did you pick the specific chart?

**Pie chart to visualize the proportion of the categories**. Here in this pie chart we are going to visualize the Proportions of the Review sentiments.

##### 2. What is/are the insight(s) found from the chart?

1. **Majority** of app reviews on the platform are **positive with 64%**, followed by **Negative** app reviews with **21.3%**,  and **neutarl reviews** with **14.7%**.

##### 3. Will the gained insights help creating a positive business impact?
Are there any insights that lead to negative growth? Justify with specific reason.

1. **High amount positive reviews-->Overall customer/user satisfaction, healthy playstore ecosystem.--> Improves app credibility**.

2. **Negative app reviews are significant, suggesting user dissatisfaction and can lead to uninstalls.**

#### Chart - 12 Multivariable analysis

In [ ]:
rev.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 29692 entries, 0 to 29691
Data columns (total 6 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   App                     29692 non-null  object 
 1   Translated_Review       29692 non-null  object 
 2   Sentiment               29692 non-null  object 
 3   Sentiment_Polarity      29692 non-null  float64
 4   Sentiment_Subjectivity  29692 non-null  float64
 5   review_len              29692 non-null  int64  
dtypes: float64(2), int64(1), object(3)
memory usage: 1.4+ MB


In [ ]:
## polarit bucket
data = rev[['Sentiment_Polarity','Sentiment_Subjectivity','review_len']]

In [ ]:
## polarit bucket and subjectivity buckets
def pol(x):
  if x<0:
    if x>-0.3 and x<0:
      return 'Low-negative-polarity'
    elif x>-0.7 and x<-0.3:
      return 'Moderate-negative-polarity'
    else:
      return 'High-negative-polarity'
  else:
    if x>0 and x<0.3:
      return 'Low-Positive-polarity'
    elif x>0.3 and x<0.7:
      return 'Moderate-positive-polarity'
    else:
      return 'High-positive-polarity'



def sub(x):
  if x==0:
    return 'objective'
  elif x>0 and x<0.3:
    return 'Low-subjectivity'
  elif x>0.3 and x<0.7:
    return 'Moderate-subjectivity'
  else:
    return 'High-subjectivity'


In [ ]:
data['pol_buc'] = data['Sentiment_Polarity'].apply(pol)
data['sub_buc'] = data['Sentiment_Subjectivity'].apply(sub)

/tmp/ipython-input-2801464501.py:1: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy

/tmp/ipython-input-2801464501.py:2: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



In [ ]:
data.head()

,Sentiment_Polarity,Sentiment_Subjectivity,review_len,pol_buc,sub_buc
0,1.00,0.533333,122,High-positive-polarity,Moderate-subjectivity
1,0.25,0.288462,47,Low-Positive-polarity,Low-subjectivity
2,0.40,0.875000,42,Moderate-positive-polarity,High-subjectivity
3,1.00,0.300000,12,High-positive-polarity,High-subjectivity
4,1.00,0.300000,8,High-positive-polarity,High-subjectivity


In [ ]:
# Chart - 12 visualization code
fig = make_subplots(
    rows=2, cols=2,
    subplot_titles=(
        'Box plot for sentiment polarity distribution wrt sentiment',
        'Box plot for sentiment subjectivtiy distribution wrt sentiment',
        'Barplot for sentiment polarity count',
        'Barplot for sentiment subjectivity count'
    )
)

# box
fig.add_trace(
    go.Box(x=rev['Sentiment'],y=rev['Sentiment_Polarity'],name='polarity vs sentiment'),
    row=1, col=1)

# box
fig.add_trace(
    go.Box(x=rev['Sentiment'],y=rev['Sentiment_Subjectivity'], name='subjectivity vs sentiment'),
    row=1, col=2)

# Barplot
df = data['pol_buc'].value_counts().reset_index()
df.sort_values(by='count',ascending=True,ignore_index=True,inplace=True)
fig.add_trace(
     go.Bar(x=df['pol_buc'],y=df['count'],name='polarity count'),
    row=2,col=1)

df1 = data['sub_buc'].value_counts().reset_index()
df1.sort_values(by='count',ascending=True,ignore_index=True,inplace=True)
fig.add_trace(
     go.Bar(x=df1['sub_buc'],y=df1['count'],name='subjectivity count'),
    row=2,col=2)

fig.update_layout(
    title='Subplots',
    title_font_size=20,
    width=1400,
    title_font_weight = 'bold',
    font= dict(size=12,weight='bold')
)


##### 1. Why did you pick the specific chart?

**Subplots to visualize multiple charts at a time and analyse multiple variables.**

##### 2. What is/are the insight(s) found from the chart?

1. **Positive reviews** have **predominantly positive polarity values** with median(0.35 moderate polarity) and **negative reviews** have **negative polarity values** with median (-0.193 low polarity). Neutral reviews have zero polarity.

2. **Positive and negative reviews show subjectivity** with medain around (0.51), indicating **users show emotions, when they like or dislike an app**. Most **Neutral reviews do not show subjectivity**, but some show subjectivity (outliers).
3. **Positive polarities** (from **Low-->High**) have high review counts **(8122(highest),7696,7552)**, indicating user satisfaction. **Low-negative-polarity** has **moderate** review count **(4186)**. **Lower review counts**(646,1489) are for **Low and moderate-negative polarites**.

4. Most reviews(16.98k) have **moderate subjectivity** suggesting **users balance opinion/emotions with facts/reasoning**. Objective and low-subjectivity reviews are very low (2k+ and 3k+).

##### 3. Will the gained insights help creating a positive business impact?
Are there any insights that lead to negative growth? Justify with specific reason.

**POSITIVES:**
1. Clear polarity and subjectivity seperation --> **Effecient Sentiment analysis**.
2. High no.of positive polarity reviews --> **user satisfaction.**

3. **High subjectivity** reviews inform about **customer experience and marketing**. Low subjective/**objective** reviews are **useful for developers**/engineering team.

**NEGATIVES**:
1. High subjectivity reviews are user emotionS--> **noise(less informative)**
2. Less negative reviews --> fewer severe issues or very **few bug detections.**

#### Chart - 13 Bivariate analysis

In [ ]:
rev.head()

,App,Translated_Review,Sentiment,Sentiment_Polarity,Sentiment_Subjectivity,review_len
0,10 Best Foods for You,I like eat delicious food. That's I'm cooking ...,Positive,1.00,0.533333,122
1,10 Best Foods for You,This help eating healthy exercise regular basis,Positive,0.25,0.288462,47
2,10 Best Foods for You,Works great especially going grocery store,Positive,0.40,0.875000,42
3,10 Best Foods for You,Best idea us,Positive,1.00,0.300000,12
4,10 Best Foods for You,Best way,Positive,1.00,0.300000,8


In [ ]:
# Chart - 13 visualization code
fig = px.scatter(
    rev,
    x='Sentiment_Polarity',
    y='Sentiment_Subjectivity',
    color='Sentiment',
    trendline='ols',
    opacity=1,
    title='Scatter plot for Sentiment polarity vs Sentiment objectivity'
)
fig.update_layout(
  title_font=dict(size=20,weight='bold'),
  font=dict(size=12,weight='bold')
)

fig.show()

##### 1. Why did you pick the specific chart?

**Scatter plot to find out the relationship between polarity and subjectivity of the reviews.**

##### 2. What is/are the insight(s) found from the chart?

1. **Most** reviews (cluster) have **polarity between -0.5 to +0.5** with **subjectivity ranging from 0.15 to 0.85**.**Indicating** reviews have **moderate sentiment** and are **balanced with user emotions and reasoning**.

2. As the **polarity moves towards extremes (-1 or +1)** the **reviews** become **subjective** (filled with user emotions), indicating **emotions increase as sentiments becomes extreme** and the **no.of reviews also decrease**(wide spread).
3. **Neutral** reviews have **zero polaity**, but have **low subjectivity**, indicating they are moslty **factual/faeture description.**

4. **Weak correlation** between subjectivity and polarity, **indicating users can express strong sentiments either emotionally or rationally.**

##### 3. Will the gained insights help creating a positive business impact?
Are there any insights that lead to negative growth? Justify with specific reason.

**POSITIVES:**
1. **Effective sentiment modelling**
2. Finding high subjective negative reviews and **addressing the issues/bugs**.
3. **Marketing** using high subjective positives reviews.

**NEGATIVES:**
* Noise due to reviews with extreme polarity and high subjectivity.

#### Chart - 14 - Correlation Heatmap

In [ ]:
data = rev[['Sentiment_Polarity','Sentiment_Subjectivity','review_len']].corr()

In [ ]:
data.head()

,Sentiment_Polarity,Sentiment_Subjectivity,review_len
Sentiment_Polarity,1.000000,0.272284,-0.154871
Sentiment_Subjectivity,0.272284,1.000000,0.030001
review_len,-0.154871,0.030001,1.000000


In [ ]:
# Correlation Heatmap visualization
fig = px.imshow(
    data,
    text_auto='.2f',
    color_continuous_scale='cividis',
    title='Correlation Heatmap',
    zmin=-1,
    zmax=1,
    aspect='auto'
)
fig.update_traces(
    textfont=dict(
        size=13,
        color='black'
    ))

fig.update_layout(
    font=dict(size=22, family='Arial Black'),
    width=1200,
    height=600,
    margin=dict(l=120, r=80, t=100, b=80)
)

fig.show()

##### 1. Why did you pick the specific chart?

**Heatmap to findout the correlation between multipile variables (Polarity,subjectivity and review length)**

##### 2. What is/are the insight(s) found from the chart?

1. **Weak positive correlation** between **subjectivity and polarity**. **No correlation** between **subjectivity and review length**.

2. **Weak Negative correlation** between **polarity and review length** i.e. **Negative reviews are slightly longer than the positive**, indicating **dissatisfied** users tend to **explain their issues in detail**.

# Merging Specific Columns from user review dataset with playstore app data

In [ ]:
rev.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 29692 entries, 0 to 29691
Data columns (total 6 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   App                     29692 non-null  object 
 1   Translated_Review       29692 non-null  object 
 2   Sentiment               29692 non-null  object 
 3   Sentiment_Polarity      29692 non-null  float64
 4   Sentiment_Subjectivity  29692 non-null  float64
 5   review_len              29692 non-null  int64  
dtypes: float64(2), int64(1), object(3)
memory usage: 1.4+ MB


In [ ]:
## creating an aggregated dataset
rev_agg = rev.groupby('App').agg({
    'review_len':'mean',
    'Sentiment_Polarity':'mean',
    'Sentiment_Subjectivity':'mean',
    'Sentiment': lambda x:x.mode()[0]
}).reset_index()

In [ ]:
rev_agg.head()

,App,review_len,Sentiment_Polarity,Sentiment_Subjectivity,Sentiment
0,10 Best Foods for You,39.936842,0.465906,0.493254,Positive
1,104 找工作 - 找工作 找打工 找兼職 履歷健檢 履歷診療室,40.861111,0.358227,0.508907,Positive
2,11st,68.384615,0.185943,0.455340,Positive
3,1800 Contacts - Lens Store,60.375000,0.318145,0.591098,Positive
4,1LINE – One Line with One Touch,22.868421,0.196290,0.557315,Positive


In [ ]:
agg = app.merge(rev_agg,on='App',how='inner') ##merging the data

In [ ]:
agg.head()

,App,Category,Rating,Reviews,Size,Installs,Type,Price,Content Rating,Genres,Last Updated,Current Ver,Android Ver,updated,year,month,primary_genre,rev_num,install_num,size_mb,price_dol,review_len,Sentiment_Polarity,Sentiment_Subjectivity,Sentiment
0,Coloring book moana,ART_AND_DESIGN,3.9,967.0,14M,"500,000+",Free,0,Everyone,Art & Design;Pretend Play,"January 15, 2018",2.0.0,4.0.3 and up,2018-01-15,2018,January,Art & Design,967.0,"500,000",14.0,0.0,13.900000,0.117917,0.645694,Positive
1,Garden Coloring Book,ART_AND_DESIGN,4.4,13791.0,33M,"1,000,000+",Free,0,Everyone,Art & Design,"September 20, 2017",2.9.2,3.0 and up,2017-09-20,2017,September,Art & Design,13791.0,"1,000,000",33.0,0.0,44.611111,0.268483,0.523447,Positive
2,FlipaClip - Cartoon animation,ART_AND_DESIGN,4.3,194216.0,39M,"5,000,000+",Free,0,Everyone,Art & Design,"August 3, 2018",2.2.5,4.0.3 and up,2018-08-03,2018,August,Art & Design,194216.0,"5,000,000",39.0,0.0,219.250000,0.412649,0.679226,Positive
3,Boys Photo Editor - Six Pack & Men's Suit,ART_AND_DESIGN,4.1,654.0,12M,"100,000+",Free,0,Everyone,Art & Design,"March 20, 2018",1.1,4.0.3 and up,2018-03-20,2018,March,Art & Design,654.0,"100,000",12.0,0.0,14.222222,0.260370,0.472593,Positive
4,Colorfit - Drawing & Coloring,ART_AND_DESIGN,4.7,20260.0,25M,"500,000+",Free,0,Everyone,Art & Design;Creativity,"October 11, 2017",1.0.8,4.0.3 and up,2017-10-11,2017,October,Art & Design,20260.0,"500,000",25.0,0.0,196.026316,0.171836,0.572762,Positive


In [ ]:
agg.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1079 entries, 0 to 1078
Data columns (total 25 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   App                     1079 non-null   object 
 1   Category                1079 non-null   object 
 2   Rating                  1078 non-null   float64
 3   Reviews                 1079 non-null   float64
 4   Size                    1079 non-null   object 
 5   Installs                1079 non-null   object 
 6   Type                    1079 non-null   object 
 7   Price                   1079 non-null   object 
 8   Content Rating          1079 non-null   object 
 9   Genres                  1079 non-null   object 
 10  Last Updated            1079 non-null   object 
 11  Current Ver             1079 non-null   object 
 12  Android Ver             1079 non-null   object 
 13  updated                 1079 non-null   object 
 14  year                    1079 non-null   

# VISUALIZATION ON THE AGGREGATED DATA


#### Chart - 15 - Bivariate analysis

In [ ]:
data = agg.groupby('Category')['Sentiment'].value_counts().reset_index()

In [ ]:
data.head()

,Category,Sentiment,count
0,ART_AND_DESIGN,Positive,10
1,AUTO_AND_VEHICLES,Positive,7
2,AUTO_AND_VEHICLES,Negative,1
3,BEAUTY,Positive,4
4,BEAUTY,Neutral,2


In [ ]:
data.sort_values(by='count',ascending=False,ignore_index=True,inplace=True)

In [ ]:
fig = px.bar(
    data,
    y='count',
    x='Category',
    text_auto=False,
    title='Sentiment of app categories',
    color='Sentiment',
    opacity=1
)

fig.update_layout(
    title_font_size=20,
    title_font_weight='bold',
    font= dict(size=12,weight='bold'),
    bargap=0.2
)

fig.show()

##### 1. Why did you pick the specific chart?

**Barchart to compare app categories wrt their review count and sentiment.**

##### 2. What is/are the insight(s) found from the chart?

1. **Games and family apps** have **higher no.of reviews**,**largely** with **positive sentiment** and some negative sentiment reviews.

2. **Neutral sentiment** reviews are observed in **categories like Tools,Photography,communication,Finance** etc.
3. **Positive sentiments dominate** across all app categories,indicating user satisfaction.
4. App categories like **comics,beauty,weather, events etc** have **very few reviews** making sentiment analysis difficult.


**POSITIVES:**
1. **Diversity** in feedbacks allows app improvement (by developers),
* Positive sentiment--> **Discoverability and scalability/growth.**
* Negative sentiment with reasoning --> **Bug identification.**

**NEGATIVES:**
1. **Sentiment volume bias towards Popular categories** and **underepresent apps with lower review count.**

# CORRELATION HEATMAP

In [ ]:
data = agg[['Rating','rev_num','price_dol','review_len','Sentiment_Polarity',	'Sentiment_Subjectivity']].corr()

In [ ]:
##heatmap code
fig = px.imshow(
    data,
    text_auto='.2f',
    color_continuous_scale='cividis',
    title='Correlation Heatmap',
    zmin=-1,
    zmax=1,
    aspect='auto'
)
fig.update_traces(
    textfont=dict(
        size=13,
        color='black'
    ))

fig.update_layout(
    font=dict(size=22, family='Arial Black'),
    width=1200,
    height=600,
    margin=dict(l=120, r=80, t=100, b=80)
)

fig.show()

INSIGHTS:
1. **Rating** is **weak positively correlated (0.22)** with **Polarity** and **subjectivity**. **Highly rated apps** have **positive sentiment** and **opinionated reviews**.

2. **Review length is negatively correlated**(moderate -0.41) with **polarity**,i.e. **Longer reviews have negative sentiment**, dissatisfied users explain issues in detail.
3. **Review count** has **positive moderate correlation (0.27)** with **Review length** i.e. **Apps with more reviews receive longer feedbacks**, therefore more **user engagement** on the platform.

## **5. Solution to Business Objective**

#### What do you suggest the client to achieve Business Objective ?
Explain Briefly.

1. **Encourage early reviews to improve ratings,rankings and discoverability**.
2. **Sentiment driven differentiation of apps**. **Negative sentiment and long reviews for bugs**, positive sentiment for ads and promotions.

3. **Content/Influencer marketing** instead of paid ads.
4. **Hybrid monetization strategy** (free apps + in app purchases+subscriptions and promotions).
5. **App category based pricing,Rank based pricing and engagement and sentiment based pricing** --> probable pricing strategy

# **Conclusion**

**The playstore ecosystem demonstrates strong growth,higher user engagegment and satisfaction as we have seen in dominant positive reviews, higher ratings and exponetial expansion post 2015.However, the growth introduces new challenges like market saturation in popular categories, Noisy feedback revenue or monetization pressure.**

**The most sustainable strategy for developers and platform is adopting Hybrid monetization strategy, early reviews, app reccomendations based on user search history and sentiment analysis, Marketing, promotions intsead of paid ads, App ranking based on user engagement and user satisfaction.**

**# Overall Transform Playstore from volume-driven-market place to user-satisfaction driven market place.**

### ***Hurrah! You have successfully completed your EDA Capstone Project !!!***